service/tablet_allocator.cc

```diff
@@ -604,6 +616,148 @@ class load_balancer {
+            load.shard_count = node->get_shard_count();
+            load.shards.resize(load.shard_count);
+            if (!load.shard_count) {
+                throw std::runtime_error(format("Shard count of {} not found in topology", host));
+            }
+        };
+        // TODO: share code with make_plan()
+        topo.for_each_node([&] (const locator::node* node_ptr) {
+            bool is_drained = node_ptr->get_state() == locator::node::state::being_decommissioned
+                              || node_ptr->get_state() == locator::node::state::being_removed;
+            if (node_ptr->get_state() == locator::node::state::normal || is_drained) {
+                ensure_node(node_ptr->host_id());
+            }
+        });
+
+        // Consider load that is already scheduled
+        for (auto&& [table, tmap_] : _tm->tablets().all_tables()) {
+            auto& tmap = *tmap_;
+            for (auto&& [tid, trinfo]: tmap.transitions()) {
+                co_await coroutine::maybe_yield();
+                if (is_streaming(&trinfo)) {
+                    auto& tinfo = tmap.get_tablet_info(tid);
+                    apply_load(nodes, get_migration_streaming_info(topo, tinfo, trinfo));
+                }
+            }
+        }
+
+        // Consider load that is about to be scheduled
+        auto& tablet_meta = _tm->tablets();
+        for (const tablet_migration_info& tmi : mplan.migrations()) {
+            co_await coroutine::maybe_yield();
+            auto& tmap = tablet_meta.get_tablet_map(tmi.tablet.table);
+            auto& tinfo = tmap.get_tablet_info(tmi.tablet.tablet);
+            auto streaming_info = get_migration_streaming_info(topo, tinfo, tmi);
+            apply_load(nodes, streaming_info);
+        }
+
+        struct repair_plan {
+            locator::global_tablet_id gid;
+            locator::tablet_info tinfo;
+            dht::token_range range;
+            dht::token last_token;
+            db_clock::duration repair_time_diff;
+        };
+
+        std::vector<repair_plan> plans;
+        for (auto&& [table, tmap_] : _tm->tablets().all_tables()) {
+            auto& tmap = *tmap_;
+            co_await coroutine::maybe_yield();
+            auto& config = tmap.repair_scheduler_config();
+            auto now = db_clock::now();
+            co_await tmap.for_each_tablet([&] (locator::tablet_id id, const locator::tablet_info& info) -> future<> {
+                auto gid = locator::global_tablet_id{table, id};
+                // Skip tablet that is in transitions.
+                auto* tti = tmap.get_tablet_transition_info(id);
+                if (tti) {
+                    lblogger.debug("Skipped tablet repair for tablet={} which is already in transition={}", gid, tti->transition);
+                    co_return;
+                }
+
+                // Skip the tablet that is about to be in transition.
+                if (co_await mplan.contains(gid)) {
+                    co_return;
```

Comment:
contains() is O(migrations_in_plan), which blows up run time of this loop by this factor. It can get significant for large clusters with lots of capacity for parallel migrations. We could make contains() O(1) by computing unordered_set<global_tablet_id> before the loop.

Reply:
OK. This is an optimization. I will calculate a set before the loop.